### ISSUE TITLE
DecimalField should work like in Python and Django

### ISSUE THREAD

**Issue by l0b0** [encode/django-rest-framework#6311]
## Checklist

- [ ] I have verified that that issue exists against the `master` branch of Django REST framework.
- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [x] This is not a usage question.
- [x] This cannot be dealt with as a third party library.
- [x] I have reduced the issue to the simplest possible case.
- [ ] I have included a failing test as a pull request.

## Steps to reproduce

```
>>> from rest_framework.serializers import DecimalField
>>> DecimalField(max_digits=40, decimal_places=20).to_internal_value('4.0')
Decimal('4.00000000000000000000')
```

## Expected behavior

The result should be `Decimal('4.0')`.

## Actual behavior

See above.

## Rationale

You might have heard the joke about the museum staff member informing the visitor that the dinosaur on display is 65 million and five years old. When the visitor, astonished, asks how they know so precisely, the staff member informs them that it's really quite simple: the dinosaur was found to be 65 million years old, and that was five years ago.

I'm working on a system which will receive engineering measurement values with various precision in text form - some are integers, others have up to at least six decimal digits of precision. It is important that these values are stored and presented with the same precision as they were recorded. Precision coercion also makes automated testing very awkward, because none of the tests look sensible when I assert that storing a specific number results in a completely different number being returned.

PostgreSQL at least does not normalize decimal numbers to a specific number of decimal places when inserting or selecting, so this should work just fine with existing databases.

As far as I can tell Django's DecimalField does the right thing:

```
>>> from django.db.models.fields import DecimalField
>>> DecimalField(max_digits=40, decimal_places=20).to_python('4.0')
Decimal('4.0')
>>> DecimalField(max_digits=40, decimal_places=20).get_prep_value('4.0')
Decimal('4.0')
```

I'm not sure if either of these are the equivalent of `to_internal_value`; please let me know if I'm checking the wrong thing.

---

Copied from issue #6149 since updates to that issue were ignored.

**Comment by carltongibson** [comment]
Hey @l0b0 — Thanks for the follow-up. "ignored" isn't quite the right word. More "slipped-through-the-gaps" I think. Sorry about that. Would have re-opened if you'd carried on the discussion there. 

I don't suppose you have a PR lined up for this do you? Thinking about a 3.9.1 now, and if you've got a fix, would be happy to review it...

**Comment by l0b0** [comment]
@carltongibson Thanks! I'll look into getting the time on the current project to do this - it would be a fairly big chunk of work to get into changing DRF internals.

**Comment by carltongibson** [comment]
> ...DRF internals...

@l0b0 It's just the one field. Comparing to Django's would probably allow you to spot the difference. If you can make some progress, I'm happy to input if you need help.

**Comment by bawn92** [comment]
I have just compared the differences between Django and the Rest Framework. The _quantize_ method in the Rest framework is applying the _decimal_place_ value passed in as an instance variable. I just want to confirm that the desired functionality is to match Django? 

In Django the DecimalField instance variables aren't applied. Is this how we would like it to function?

```
>>> from django.db.models.fields import DecimalField
>>> DecimalField(max_digits=5, decimal_places=30).to_python('4.00000000')
Decimal('4.00000000')
```

**Comment by l0b0** [comment]
@bawn92 I believe `max_digits` and `decimal_places` are meant to be *limits* rather than a format. That way it keeps the number of significant digits in the normal case, which is what I would expect from a reasonable implementation.

**Comment by tomchristie** [comment]
I'm not sure this is super obvious either way around. Obviously it's desirable in some cases, but also, would we expect "4" or "4.0"? Also for currencies, you'd typically want to a strict number of decimal places, eg `"21.50"`, similarly with some time resolutions it might be desirable to strictly output a fixed number of decimal places.

> It is important that these values are stored and presented with the same precision as they were recorded

It's possible that you want a custom field, since that's actually quite precise. Eg. Do you want `1.0` and `1.000` to be stored differently in the database?

**Comment by l0b0** [comment]
@tomchristie The point is to keep the number of significant digits - have a look at the anecdote from the original report. I know that's a joke, but it highlights that precision (as opposed to accuracy) is important in science (and, in my case, engineering). `Decimal('4') == Decimal('4.0')`, but the former implies a precision of ±0.5 while the latter a precision of ±0.05. This makes a massive difference for how information should be displayed. If I have to do any kind of non-trivial arithmetic with a number like `Decimal('1.2345')` then I know with 100% certainty that the displaying the result with two digits of precision and with 20 digits of precision are both wrong - the former loses precision which was present in the input data, and the latter implies precision which was not present in the input data.

But all this is academic. Python doesn't bork the precision, Postgres doesn't bork the precision, and Django doesn't bork the precision. So I think it's reasonable that DRF shouldn't bork the precision.

**Comment by tomchristie** [comment]
My point is, should the significant digits be preserved based on:

* The input given by the user. Treat "5" distinctly from "5.0", or "5.000", and hold on to the *input* precision. You've got some extra "what was the input precision" information that needs to be stored in addition to the raw number itself.
* Assume that trailing zeros are insignificant (which they may or may not *actually* be.) Always treat five and a half as "5.5" Always treat five as "5" or "5.0". (Which of those two?) In currencies, or athletic time precisions, that's probably not what you want.
* Assume that the precision used in the data modeling is the precision you want to see represented. Treat five as being properly displayed with whatever precision the data modeling has been set to. This is what we do at the moment.

The issue you're seeing is when you've got a case where the precision of the data modeling (20 digits) does not match the precision you *actually* want to see.

I'd suggest that we've probably got two viable candidates for display choices, both of which are relevant. If we added a `strip_trailing_zeros=False` flag, then users would be able to switch between "10.000" and "10.0" displays.

Couple of examples of use cases.

* Currency: We'd always like to display "10.00". User input of "10" still means precisely "10.00".
* Known-precision timing: If we have (say) millisecond precision timer then we always wan to display a fixed number of digits "10.000" really does represent the precision we meant.
* Variable data: We'd like to treat the input as a decimal, but don't know what precision it will be in. (The data modeling just specifies the maximum possible precision that's allowed) Display "10.0".

> I believe [...] decimal_places are meant to be limits rather than a format.

As above, there are different contexts, and could mean either *exactly this precision*, or *anything up to this precision* in different cases.

**Comment by tomchristie** [comment]
Happy to add a `strip_trailing_zeros` flag - will accept PRs for that.
